File: src/main/java/me/zeroeightsix/kami/module/modules/player/Scaffold.java
package me.zeroeightsix.kami.module.modules.player;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.EntityUtil;
import me.zeroeightsix.kami.util.Wrapper;
import net.minecraft.block.Block;
import net.minecraft.block.BlockContainer;
import net.minecraft.block.BlockFalling;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.network.play.client.CPacketEntityAction;
import net.minecraft.network.play.client.CPacketEntityAction.Action;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3d;
import net.minecraftforge.client.event.InputUpdateEvent;

import static me.zeroeightsix.kami.KamiMod.MODULE_MANAGER;
import static me.zeroeightsix.kami.util.BlockInteractionHelper.*;

/**
 * Created by 086 on 20/01/19
 * Updated by Polymer on 16/01/20
 * Updated by S-B99 on 02/03/20
 * @see me.zeroeightsix.kami.mixin.client.MixinEntity
 */
@Module.Info(name = "Scaffold", category = Module.Category.PLAYER, description = "Places blocks under you")
public class Scaffold extends Module {

    private Setting<Boolean> placeBlocks = register(Settings.b("Place Blocks", true));
    private Setting<Mode> modeSetting = register(Settings.enumBuilder(Mode.class).withName("Mode").withValue(Mode.LEGIT).build());
    private Setting<Boolean> randomDelay = register(Settings.booleanBuilder("Random Delay").withValue(false).withVisibility(v -> modeSetting.getValue().equals(Mode.LEGIT)).build());
    private Setting<Integer> delayRange = register(Settings.integerBuilder("Delay Range").withMinimum(0).withValue(6).withMaximum(10).withVisibility(v -> modeSetting.getValue().equals(Mode.LEGIT) && randomDelay.getValue()).build());
    private Setting<Integer> ticks = register(Settings.integerBuilder("Ticks").withMinimum(0).withMaximum(60).withValue(2).withVisibility(v -> !modeSetting.getValue().equals(Mode.LEGIT)).build());

    private boolean shouldSlow = false;

    private static Scaffold INSTANCE;

    public Scaffold() {
        INSTANCE = this;
    }

    public static boolean shouldScaffold() {
        return INSTANCE.isEnabled();
    }

    private enum Mode {
        NEITHER, LEGIT
    }

    @EventHandler
    private Listener<InputUpdateEvent> eventListener = new Listener<>(event -> {
        if (modeSetting.getValue().equals(Mode.LEGIT) && shouldSlow) {
            if (randomDelay.getValue()) {
                event.getMovementInput().moveStrafe *= 0.2f + getRandomInRange();
                event.getMovementInput().moveForward *= 0.2f + getRandomInRange();
            }
            else {
                event.getMovementInput().moveStrafe *= 0.2f;
                event.getMovementInput().moveForward *= 0.2f;
            }
        }
    });

    @Override
    public void onUpdate() {
        if (mc.player == null || MODULE_MANAGER.isModuleEnabled(Freecam.class)) return;
        shouldSlow = false;

        Vec3d vec3d = EntityUtil.getInterpolatedPos(mc.player, ticks.getValue());
        if (modeSetting.getValue().equals(Mode.LEGIT)) vec3d = EntityUtil.getInterpolatedPos(mc.player, 0);

        BlockPos blockPos = new BlockPos(vec3d).down();
        BlockPos belowBlockPos = blockPos.down();
        BlockPos legitPos = new BlockPos(EntityUtil.getInterpolatedPos(mc.player, 2));

        /* when legitBridge is enabled */
        /* check if block behind player is air or other replaceable block and if it is, make the player crouch */
        if (modeSetting.getValue().equals(Mode.LEGIT) && Wrapper.getWorld().getBlockState(legitPos.down()).getMaterial().isReplaceable() && mc.player.onGround) {
            shouldSlow = true;
            mc.player.movementInput.sneak = true;
            mc.player.connection.sendPacket(new CPacketEntityAction(mc.player, Action.START_SNEAKING));
        }

        /* check if block is already placed */
        if (!Wrapper.getWorld().getBlockState(blockPos).getMaterial().isReplaceable()) {
            return;
        }

        setSlotToBlocks(belowBlockPos);

        /* check if we don't have a block adjacent to the blockPos */
        if (!checkForNeighbours(blockPos)) return;

        /* place the block */
        if (placeBlocks.getValue()) placeBlockScaffold(blockPos);
        mc.player.connection.sendPacket(new CPacketEntityAction(mc.player, Action.STOP_SNEAKING));
        shouldSlow = false;
    }

    private float getRandomInRange() {
        return 0.11f + (float) Math.random() * ((delayRange.getValue() / 10.0f) - 0.11f);
    }

    private void setSlotToBlocks(BlockPos belowBlockPos) {
        /* search blocks in hotbar */
        int newSlot = -1;
        for (int i = 0; i < 9; i++) {
            /* filter out non-block items */
            ItemStack stack = Wrapper.getPlayer().inventory.getStackInSlot(i);
            if (stack == ItemStack.EMPTY || !(stack.getItem() instanceof ItemBlock)) continue;

            Block block = ((ItemBlock) stack.getItem()).getBlock();
            if (blackList.contains(block) || block instanceof BlockContainer) continue;

            /* filter out non-solid blocks */
            if (!Block.getBlockFromItem(stack.getItem()).getDefaultState().isFullBlock()) continue;

            /* don't use falling blocks if it'd fall */
            if (((ItemBlock) stack.getItem()).getBlock() instanceof BlockFalling) {
                if (Wrapper.getWorld().getBlockState(belowBlockPos).getMaterial().isReplaceable()) continue;
            }
            newSlot = i;
            break;
        }
        /* check if any blocks were found, and if they were then set the slot */
        int oldSlot = 1; /* make it 1, instead of -1 so you don't get kicked if it was -1 */
        if (newSlot != -1) {
            oldSlot = Wrapper.getPlayer().inventory.currentItem;
            Wrapper.getPlayer().inventory.currentItem = newSlot;
        }
        /* reset slot back to the original one */
        Wrapper.getPlayer().inventory.currentItem = oldSlot;
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/player/Scaffold.kt
package me.zeroeightsix.kami.module.modules.player

import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import me.zeroeightsix.kami.event.Phase
import me.zeroeightsix.kami.event.SafeClientEvent
import me.zeroeightsix.kami.event.events.OnUpdateWalkingPlayerEvent
import me.zeroeightsix.kami.event.events.PacketEvent
import me.zeroeightsix.kami.event.events.PlayerTravelEvent
import me.zeroeightsix.kami.manager.managers.PlayerPacketManager
import me.zeroeightsix.kami.mixin.client.entity.MixinEntity
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import me.zeroeightsix.kami.util.*
import me.zeroeightsix.kami.util.EntityUtils.prevPosVector
import me.zeroeightsix.kami.util.WorldUtils.placeBlock
import me.zeroeightsix.kami.util.math.RotationUtils
import me.zeroeightsix.kami.util.math.Vec2f
import me.zeroeightsix.kami.util.math.VectorUtils.toBlockPos
import me.zeroeightsix.kami.util.threads.defaultScope
import me.zeroeightsix.kami.util.threads.onMainThreadSafe
import me.zeroeightsix.kami.util.threads.safeListener
import net.minecraft.item.ItemBlock
import net.minecraft.network.play.client.CPacketEntityAction
import net.minecraft.network.play.server.SPacketPlayerPosLook
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.RayTraceResult
import net.minecraft.util.math.Vec3d
import org.kamiblue.event.listener.listener
import kotlin.math.floor
import kotlin.math.roundToInt

/**
 * @see MixinEntity.isSneaking
 */
object Scaffold : Module(
    name = "Scaffold",
    category = Category.PLAYER,
    description = "Places blocks under you",
    modulePriority = 500
) {
    private val tower by setting("Tower", true)
    private val spoofHotbar by setting("SpoofHotbar", true)
    val safeWalk by setting("SafeWalk", true)
    private val sneak by setting("Sneak", true)
    private val delay by setting("Delay", 2, 1..10, 1)
    private val maxRange by setting("MaxRange", 1, 0..3, 1)

    private var lastRotation = Vec2f.ZERO
    private var placeInfo: Pair<EnumFacing, BlockPos>? = null
    private var inactiveTicks = 69

    private val placeTimer = TickTimer(TimeUnit.TICKS)
    private val rubberBandTimer = TickTimer(TimeUnit.TICKS)

    override fun isActive(): Boolean {
        return isEnabled && inactiveTicks <= 5
    }

    init {
        onDisable {
            placeInfo = null
            inactiveTicks = 69
        }

        listener<PacketEvent.Receive> {
            if (it.packet !is SPacketPlayerPosLook) return@listener
            rubberBandTimer.reset()
        }

        safeListener<PlayerTravelEvent> {
            if (!tower || !mc.gameSettings.keyBindJump.isKeyDown || inactiveTicks > 5 || !isHoldingBlock) return@safeListener
            if (rubberBandTimer.tick(10, false)) {
                if (shouldTower) player.motionY = 0.41999998688697815
            } else if (player.fallDistance <= 2.0f) {
                player.motionY = -0.169
            }
        }
    }

    private val isHoldingBlock: Boolean
        get() = PlayerPacketManager.getHoldingItemStack().item is ItemBlock

    private val SafeClientEvent.shouldTower: Boolean
        get() = !player.onGround
            && player.posY - floor(player.posY) <= 0.1
    init {
        safeListener<OnUpdateWalkingPlayerEvent> { event ->
            if (event.phase != Phase.PRE) return@safeListener

            inactiveTicks++
            placeInfo = calcNextPos()?.let {
                WorldUtils.getNeighbour(it, 1, sides = arrayOf(EnumFacing.DOWN))
                    ?: WorldUtils.getNeighbour(it, 3, sides = EnumFacing.HORIZONTALS)
            }

            placeInfo?.let {
                val hitVec = WorldUtils.getHitVec(it.second, it.first)
                lastRotation = RotationUtils.getRotationTo(hitVec)
                swapAndPlace(it.second, it.first)
            }

            if (inactiveTicks > 5) {
                PlayerPacketManager.resetHotbar()
            } else if (isHoldingBlock) {
                val packet = PlayerPacketManager.PlayerPacket(rotating = true, rotation = lastRotation)
                PlayerPacketManager.addPacket(this@Scaffold, packet)
            }
        }
    }

    private fun SafeClientEvent.calcNextPos(): BlockPos? {
        val posVec = player.positionVector
        val blockPos = posVec.toBlockPos()
        return checkPos(blockPos)
            ?: run {
                val realMotion = posVec.subtract(player.prevPosVector)
                val nextPos = blockPos.add(roundToRange(realMotion.x), 0, roundToRange(realMotion.z))
                checkPos(nextPos)
            }
    }

    private fun SafeClientEvent.checkPos(blockPos: BlockPos): BlockPos? {
        val center = Vec3d(blockPos.x + 0.5, blockPos.y.toDouble(), blockPos.z + 0.5)
        val rayTraceResult = world.rayTraceBlocks(
            center,
            center.subtract(0.0, 0.5, 0.0),
            false,
            true,
            false
        )
        return blockPos.down().takeIf { rayTraceResult?.typeOfHit != RayTraceResult.Type.BLOCK }
    }

    private fun roundToRange(value: Double) =
        (value * 2.5 * maxRange).roundToInt().coerceAtMost(maxRange)

    private fun SafeClientEvent.swapAndPlace(pos: BlockPos, side: EnumFacing) {
        getBlockSlot()?.let { slot ->
            if (spoofHotbar) PlayerPacketManager.spoofHotbar(slot)
            else InventoryUtils.swapSlot(slot)

            inactiveTicks = 0

            if (placeTimer.tick(delay.toLong())) {
                val shouldSneak = sneak && !player.isSneaking
                defaultScope.launch {
                    if (shouldSneak) {
                        player.let {
                            it.connection.sendPacket(CPacketEntityAction(it, CPacketEntityAction.Action.START_SNEAKING))
                        }
                    }
                    delay(5)
                    onMainThreadSafe {
                        placeBlock(pos, side)
                        if (shouldSneak) {
                            connection.sendPacket(CPacketEntityAction(player, CPacketEntityAction.Action.STOP_SNEAKING))
                        }
                    }
                }
            }
        }
    }

    private fun SafeClientEvent.getBlockSlot(): Int? {
        playerController.updateController()
        for (i in 0..8) {
            val itemStack = player.inventory.mainInventory[i]
            if (itemStack.isEmpty || itemStack.item !is ItemBlock) continue
            return i
        }
        return null
    }

}